sdio/imx6ull/imx6ull-sdio.c

```diff
@@ -0,0 +1,855 @@
+
+	rslt = sdio_cmdSend(SDIO_CMD_RW_DIRECT, arg, &response);
+
+	/* Data is part of response on read transfers
+	 * and also on writes with RaW (read after write)
+	 * flag set - SDIO simplified spec v3.00, page 36 */
+	if (dir == sdio_read) {
+		*data = response & 0xff;
+	}
+
+	return rslt;
+}
+
+
+static int _sdio_transferBulk(sdio_dir_t dir, int blockMode, uint32_t address,
+	uint8_t area, uint8_t *data, size_t len)
+{
+	uint32_t val, cmd, mix, blk;
+	uint32_t arg = 0;
+	uint16_t count;
+
+	if (blockMode == 1 && (sdio_common.blocksz == 0 || len % sdio_common.blocksz != 0)) {
+		printf("libsdio: transfer size is not a multiple of block size\n");
+		return EINVAL;
+	}
+	count = (blockMode) ? (len / sdio_common.blocksz) : len;
+
+	/* construct argument */
+	arg |= (dir << 31);
+	arg |= ((area & 0x7) << 28);
+	arg |= ((address & 0x1ffff) << 9);
+	arg |= (count & 0x1ff);
+	arg |= ((blockMode & 0x1) << 27);
+	arg |= (1 << 26);
+
+	val = *(sdio_common.base + reg_pres_state);
+	if (val & 0x7) {
+		printf("libsdio: data interface is busy (pres_state=%x)\n", val);
+		return -1;
+	}
+
+	val = *(sdio_common.base + reg_int_status);
+	if (val & 0x2) {
+		printf("libsdio: transfer has been aborted (int_status=%x)\n", val);
+		return -1;
+	}
+
+	if (len > DMA_BUFFER_SIZE) {
+		printf("libsdio: transfer size is too large (length=%u > %u)\n", len, DMA_BUFFER_SIZE);
+		return -1;
+	}
+
+	if (dir == sdio_write) {
+		memcpy(sdio_common.dmaptr, data, len);
+	}
+
+	cmd = SDIO_CMD_RW_EXTENDED << 24;
+	cmd |= 1 << 21;   /* DPSEL=1 */
+	cmd |= 1 << 20;   /* CICEN=1 */
+	cmd |= 1 << 19;   /* CCCEN=1 */
+	cmd |= 0x2 << 16; /* RSPTYP=2 */
+
+	mix = (1U << 31) | (1 << 0); /* DMAEN=1 */
+	if (dir == sdio_read) {
+		mix |= 1 << 4; /* DTDSEL=1 */
+	}
+
+	if (blockMode == 1) {
+		mix |= (1 << 5); /* MSBSEL=1 */
+		mix |= (1 << 1); /* BCEN=1 */
+
+		blk = (count << 16) | sdio_common.blocksz;
+	}
+	else {
+		blk = (1 << 16) | len;
+	}
+
+	*(sdio_common.base + reg_mix_ctrl) = mix;
+	*(sdio_common.base + reg_blk_att) = blk;
+	*(sdio_common.base + reg_ds_addr) = sdio_common.dmaphys;
+	*(sdio_common.base + reg_cmd_arg) = arg;
+	*(sdio_common.base + reg_cmd_xfer_typ) = cmd;
+
+	/* wait 1 ms max */
+	if (sdhc_cmdWait(0xb, 10, 100) < 0) {
+		sdhc_reset(SDHC_SYSCTL_RESET_CMD);
+		sdhc_reset(SDHC_SYSCTL_RESET_DATA);
+		return -2;
+	}
+
+	/* clear status flags */
+	*(sdio_common.base + reg_int_status) = 0xb; /* DINT=1 TC=1 CC=1 */
+
+	if (dir == sdio_read) {
+		memcpy(data, sdio_common.dmaptr, len);
+	}
+
+	/* retrieve response */
+	val = *(sdio_common.base + reg_cmd_rsp0);
+
+	return 0;
+}
+
+
+int sdio_transferBulk(sdio_dir_t dir, int blockMode, uint32_t address,
+	uint8_t area, uint8_t *data, size_t len)
+{
+	int rslt;
+
+	mutexLock(sdio_common.cmdLock);
+	rslt = _sdio_transferBulk(dir, blockMode, address, area, data, len);
+	mutexUnlock(sdio_common.cmdLock);
+
+	return rslt;
+}
+
+
+void sdio_eventRegister(uint8_t event, sdio_event_handler_t handler, void *arg)
+{
+	if (event > 18) {
```

Comment:
Add a define for `19`, which is the number of possible events.
```suggestion
	if (event >= N_EVENTS) {
```

Reply:
Applied as suggested.